File: sdk/python/x402/verify.py
"""Signature verification for x402 payments.

This module uses the native Rust implementation via FFI when available,
falling back to pure Python for compatibility.
"""

import time
from typing import Optional

from eth_account.messages import encode_defunct
from eth_account import Account

from x402.types import PaymentRequirements, SignedPayment
from x402.protocol import decode_payment_header

# Try to import native Rust bindings
try:
    from x402_native import (
        verify_signed_payment as _native_verify,
        PaymentRequirements as NativePaymentRequirements,
    )
    _USE_NATIVE = True
except ImportError:
    _USE_NATIVE = False


class X402VerificationError(Exception):
    """Base exception for verification errors."""
    pass


class PaymentExpiredError(X402VerificationError):
    """Payment has expired."""
    pass


class InsufficientAmountError(X402VerificationError):
    """Payment amount is insufficient."""
    
    def __init__(self, required: int, provided: int):
        self.required = required
        self.provided = provided
        super().__init__(f"Insufficient amount: required {required}, got {provided}")


class InvalidSignatureError(X402VerificationError):
    """Signature is invalid."""
    pass


def verify_payment(
    payment_header: str,
    requirements: PaymentRequirements,
    current_time: Optional[int] = None,
) -> str:
    """Verify a signed payment against requirements.
    
    Uses native Rust implementation when available.
    
    Args:
        payment_header: The X-Payment header value
        requirements: Payment requirements to verify against
        current_time: Current unix timestamp (defaults to now)
        
    Returns:
        The verified payer address
        
    Raises:
        PaymentExpiredError: If payment has expired
        InsufficientAmountError: If amount is insufficient
        InvalidSignatureError: If signature is invalid
    """
    if _USE_NATIVE:
        # Use Rust implementation for verification
        try:
            native_req = NativePaymentRequirements(
                amount=requirements.amount,
                recipient=requirements.recipient,
                network=requirements.network.value if hasattr(requirements.network, 'value') else str(requirements.network),
                resource=requirements.resource,
                token=requirements.token,
                description=requirements.description,
                expires_at=requirements.expires_at,
            )
            return _native_verify(payment_header, native_req)
        except ValueError as e:
            error_msg = str(e)
            if "expired" in error_msg.lower():
                raise PaymentExpiredError(error_msg)
            elif "insufficient" in error_msg.lower():
                raise InsufficientAmountError(0, 0)  # TODO: parse amounts
            else:
                raise InvalidSignatureError(error_msg)
    
    # Fallback: pure Python implementation
    # Decode the payment
    signed_payment = decode_payment_header(payment_header)
    payment = signed_payment.payment
    
    # Check expiry
    now = current_time or int(time.time())
    if payment.expires_at < now:
        raise PaymentExpiredError("Payment has expired")
    
    # Check amount
    if payment.amount < requirements.amount:
        raise InsufficientAmountError(requirements.amount, payment.amount)
    
    # Check recipient
    if payment.recipient.lower() != requirements.recipient.lower():
        raise InvalidSignatureError(
            f"Recipient mismatch: expected {requirements.recipient}, got {payment.recipient}"
        )
    
    # Check network
    expected_chain_id = requirements.network.chain_id if hasattr(requirements.network, 'chain_id') else _get_chain_id(requirements.network)
    if payment.chain_id != expected_chain_id:
        raise InvalidSignatureError(
            f"Chain ID mismatch: expected {expected_chain_id}, got {payment.chain_id}"
        )
    
    # Verify signature and recover address
    message_hash = payment.message_hash()
    
    try:
        recovered_address = Account.recover_message(
            encode_defunct(primitive=message_hash),
            signature=signed_payment.signature
        )
    except Exception as e:
        raise InvalidSignatureError(f"Failed to recover address: {e}")
    
    # Check recovered address matches payer
    if recovered_address.lower() != payment.payer.lower():
        raise InvalidSignatureError(
            f"Recovered address {recovered_address} does not match payer {payment.payer}"
        )
    
    return recovered_address


def _get_chain_id(network: str) -> int:
    """Get chain ID from network string."""
    chain_ids = {
        "ethereum": 1,
        "base": 8453,
        "base_sepolia": 84532,
        "arbitrum": 42161,
        "optimism": 10,
        "polygon": 137,
    }
    return chain_ids.get(network, 0)

File: core/src/verify.rs
//! Signature verification for x402 payments

use crate::{SignedPayment, PaymentRequirements, X402Error, Result};
use alloy_primitives::Address;
use k256::ecdsa::{RecoveryId, Signature, VerifyingKey};

/// Verify a signed payment against requirements
/// 
/// Checks:
/// 1. Signature is valid and recovers to payer address
/// 2. Amount meets requirements
/// 3. Recipient matches
/// 4. Payment not expired
/// 5. Network matches
pub fn verify_payment(
    payment: &SignedPayment,
    requirements: &PaymentRequirements,
) -> Result<Address> {
    // Check expiry
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_secs();
    
    if payment.payment.expires_at < now {
        return Err(X402Error::PaymentExpired);
    }

    // Check amount (convert U256 to u64 for comparison - simplified)
    let required_amount: u64 = requirements.amount.try_into()
        .unwrap_or(u64::MAX);
    let provided_amount: u64 = payment.payment.amount.try_into()
        .unwrap_or(0);
    
    if provided_amount < required_amount {
        return Err(X402Error::InsufficientAmount {
            required: required_amount,
            provided: provided_amount,
        });
    }

    // Check recipient
    if payment.payment.recipient != requirements.recipient {
        return Err(X402Error::InvalidSignature(
            "recipient mismatch".to_string()
        ));
    }

    // Check network
    if payment.payment.chain_id != requirements.network.chain_id() {
        return Err(X402Error::UnsupportedNetwork(
            format!("expected chain {}, got {}", 
                requirements.network.chain_id(), 
                payment.payment.chain_id)
        ));
    }

    // Verify signature and recover payer address
    let recovered_address = recover_signer(payment)?;
    
    if recovered_address != payment.payment.payer {
        return Err(X402Error::InvalidSignature(
            "recovered address does not match payer".to_string()
        ));
    }

    Ok(recovered_address)
}

/// Recover the signer address from a signed payment
pub fn recover_signer(payment: &SignedPayment) -> Result<Address> {
    if payment.signature.len() != 65 {
        return Err(X402Error::InvalidSignature(
            format!("signature must be 65 bytes, got {}", payment.signature.len())
        ));
    }

    let message_hash = payment.payment.message_hash();
    
    // Parse signature components
    let r_s = &payment.signature[..64];
    let v = payment.signature[64];
    
    // Recovery ID: v is either 27/28 (legacy) or 0/1
    let recovery_id = if v >= 27 {
        RecoveryId::try_from(v - 27)
    } else {
        RecoveryId::try_from(v)
    }.map_err(|_| X402Error::InvalidSignature("invalid recovery id".to_string()))?;

    let signature = Signature::from_slice(r_s)
        .map_err(|e| X402Error::InvalidSignature(e.to_string()))?;

    let verifying_key = VerifyingKey::recover_from_prehash(&message_hash, &signature, recovery_id)
        .map_err(|e| X402Error::InvalidSignature(e.to_string()))?;

    // Convert public key to Ethereum address
    let public_key_bytes = verifying_key.to_encoded_point(false);
    let public_key_hash = alloy_primitives::keccak256(&public_key_bytes.as_bytes()[1..]);
    let address = Address::from_slice(&public_key_hash[12..]);

    Ok(address)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_invalid_signature_length() {
        use crate::{PaymentPayload, SignedPayment};
        use alloy_primitives::U256;

        let payment = SignedPayment {
            payment: PaymentPayload {
                amount: U256::from(1000),
                recipient: Address::ZERO,
                payer: Address::ZERO,
                chain_id: 8453,
                token: None,
                resource: "/test".to_string(),
                nonce: 1,
                expires_at: u64::MAX,
            },
            signature: vec![0u8; 64], // Wrong length
        };

        let result = recover_signer(&payment);
        assert!(result.is_err());
    }
}
